tests/test_converter_cqasm.py

```diff
@@ -0,0 +1,255 @@
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# As a special exception, the copyright holders of exqalibur library give you
+# permission to combine exqalibur with code included in the standard release of
+# Perceval under the MIT license (or modified versions of such code). You may
+# copy and distribute such a combined system following the terms of the MIT
+# license for both exqalibur and Perceval. This exception for the usage of
+# exqalibur is limited to the python bindings used by Perceval.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import pytest
+
+from pathlib import Path
+
+from perceval import StateVector, pdisplay
+from perceval.converters import CQASMConverter, ConversionSyntaxError, ConversionUnsupportedFeatureError
+import perceval.components.unitary_components as components
+from perceval.components import catalog
+from perceval.rendering.format import Format
+
+import numpy as np
+
+
+def test_converter_syntax_error():
+    source = """
+version 3
+rabbit r
+"""
+    with pytest.raises(ConversionSyntaxError):
+        CQASMConverter(catalog).convert_string(source)
+
+
+def test_converter_unsupported_classical_variable():
+    source = """
+version 3
+qubit[2] q
+bit b
+X q[0]
+"""
+    with pytest.raises(ConversionUnsupportedFeatureError):
+        CQASMConverter(catalog).convert_string(source)
+
+
+def test_converter_unsupported_gates():
+    source = """
+version 3
+qubit[3] q
+CNOT q[0:2], q[2]
+"""
+    # Caught early: two controls for one target
+    with pytest.raises(ConversionUnsupportedFeatureError):
+        CQASMConverter(catalog).convert_string(source)
+
+    source = """
+version 3
+qubit[2] q
+CR(pi) q[0], q[1]
+"""
+    # Caught later in the abstract class
+    with pytest.raises(ConversionUnsupportedFeatureError):
+        CQASMConverter(catalog).convert_string(source)
+
+
+def test_converter_bell_state():
+    source = """
+version 3
+qubit[2] q
+H q[0]
+CNOT q[0], q[1]
+"""
+    pc = CQASMConverter(catalog).convert_string(source, use_postselection=False)
+    assert pc.circuit_size == 8
+    assert pc.m == 4
+    assert pc.source_distribution[StateVector('|1,0,1,0,0,1,0,1>')] == 1
+    assert len(pc._components) == 2  # should be  BS.H // CNOT
```

Comment:
Please add an assert to check `_components` in the converted circuit `pc` are the ones in QASM input. Like in this test to check that they are `H` and `CNOT`.

Reply:
Sure, done for all tests!